cpp/src/arrow/CMakeLists.txt

```diff
@@ -244,28 +243,7 @@ foreach(LIB_TARGET ${ARROW_LIBRARIES})
 endforeach()
 
 # Headers: top level
-install(FILES
-  allocator.h
-  api.h
-  array.h
-  buffer.h
-  builder.h
-  compare.h
-  memory_pool.h
-  pretty_print.h
-  record_batch.h
-  status.h
-  stl.h
-  table.h
-  table_builder.h
-  tensor.h
-  type.h
-  type_fwd.h
-  type_traits.h
-  test-util.h
-  visitor.h
-  visitor_inline.h
-  DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/arrow")
+ARROW_INSTALL_ALL_HEADERS("arrow")
```

Comment:
IIRC in some cases we don't want to install all targets (some of them are private). Though I don't think we have been very consistent about this.

Reply:
This is true. Perhaps we can come up with a naming convention for private headers so that they can be automatically excluded by this function

I opened https://issues.apache.org/jira/browse/ARROW-4029